uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestComment.php

```diff
@@ -0,0 +1,127 @@
+
+class CRM_HRLeaveAndAbsences_Service_LeaveRequestComment {
+
+  /**
+   * @var string
+   */
+  private $entityName = 'LeaveRequest';
+
+  /**
+   * @var array
+   *   The parameter array needed to add, create or delete comments
+   */
+  private $params;
+
+  /**
+   * CRM_HRLeaveAndAbsences_Service_LeaveRequestComment constructor.
+   *
+   * @param $params
+   *   The parameter array needed to add, create or delete comments
+   */
+  public function __construct($params) {
+    $this->params = $params;
+    $this->prepareParametersForCommentPayload();
+  }
+
+  /**
+   * Uses the API exposed by the Comment entity (Comment.create)
+   * to create comments related to a LeaveRequest.
+   *
+   * @throws UnexpectedValueException
+   *
+   * @return array
+   */
+  public function add() {
+    //updates are not allowed for now
+    if (!empty($this->params['id'])){
+      throw new UnexpectedValueException('You cannot update a comment!');
+    }
+
+    $results =  civicrm_api3('Comment', 'create', $this->params);
+
+    if($results['count'] > 0) {
+      $this->filterCommentReturnData($results['values']);
+    }
+
+    return $results;
+  }
+
+  /**
+   * Uses the API exposed by the Comment entity (Comment.get)
+   * to fetch comments associated with a LeaveRequest
+   *
+   * @return array
+   */
+  public function get() {
+    $results =  civicrm_api3('Comment', 'get', $this->params);
+
+    if($results['count'] > 0) {
+      $this->filterCommentReturnData($results['values']);
+    }
+
+    return $results;
+  }
+
+  /**
+   * Uses the API exposed by the Comment entity (Comment.delete)
+   * to delete comments associated with a LeaveRequest.
+   * This method also implement some checks to ensure that only the LeaveRequest Approver
+   * or an Admin can delete a comment
+   *
+   * @throws UnexpectedValueException
+   * @throws InvalidArgumentException
+   *
+   * @return array
+   */
+  public function delete() {
+    $comments =  civicrm_api3('Comment', 'get', ['id' => $this->params['id'], 'sequential' => 1]);
+
+    if ($comments['count'] > 0) {
+      $leaveRequest = LeaveRequest::findById($comments['values'][0]['entity_id']);
+      $leaveManagerService = new LeaveManagerService();
+
+      if ($leaveManagerService->currentUserIsAdmin() || $leaveManagerService->currentUserIsLeaveManagerOf($leaveRequest->contact_id)) {
+        return civicrm_api3('Comment', 'delete', $this->params);
+      }
+
+      throw new UnexpectedValueException('You must either be an L&A admin or an approver to this leave request to be able to delete the comment');
+    }
+
+    throw new InvalidArgumentException('Comment does not exist or has been deleted already!');
+  }
+
+  /**
+   * Helper function used to format the parameters
+   * into a format expected by the Comment.create, Comment.delete and Comment.get API
+   */
+  private function prepareParametersForCommentPayload() {
+    $this->params['entity_name'] = $this->entityName;
```

Comment:
With the removal of `$params` from the constructor, I think this method will need to be changed a bit. I would create a `callCommentAPI` method:

```php
private function callCommentAPI($action, $params) {
   $params = $this->prepareParametersForCommentPayload($params);
   
   return civicrm_api3('Comment', $action, $params);
}
```

And then, would call this method inside the other ones.